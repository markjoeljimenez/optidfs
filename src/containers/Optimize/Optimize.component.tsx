import { useAppDispatch, useAppSelector } from '../../hooks';
import {
	setActiveStackingTab,
	STACKING_TYPE,
} from '../Stacking/Stacking.actions';
import { setActiveTab } from '../Tabs/Tabs.actions';
import { optimize } from './Optimize.actions';

interface IOptimizeProps {
	disabled?: boolean;
}

const Optimize = ({ disabled }: IOptimizeProps) => {
	const dispatch = useAppDispatch();
	// TODO: Undo this
	// const { stacking } = useAppSelector((state) => state);
	const stacking = {} as any;

	function handleClick() {
		if (
			stacking &&
			stacking.TEAM &&
			!stacking.TEAM?.NUMBER_OF_PLAYERS_TO_STACK
		) {
			dispatch(setActiveTab('stacking'));
			dispatch(setActiveStackingTab(STACKING_TYPE.TEAM));

			return;
		}

		if (
			stacking &&
			stacking.POSITION &&
			!stacking.POSITION?.NUMBER_OF_POSITIONS
		) {
			dispatch(setActiveTab('stacking'));
			dispatch(setActiveStackingTab(STACKING_TYPE.POSITION));

			return;
		}

		dispatch(optimize());
		dispatch(setActiveTab('players'));
	}

	return (
		<button
			className="px-4 py-3 bg-indigo-700 text-white rounded shadow font-black hover:bg-indigo-800"
			disabled={disabled}
			type="button"
			onClick={handleClick}
		>
			Optimize
		</button>
	);
};

export default Optimize;
