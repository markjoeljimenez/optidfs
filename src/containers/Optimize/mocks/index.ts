export * from './getOptimizedLineupsPostBody.mocks';
export * from './optimizedLineups.mocks';
