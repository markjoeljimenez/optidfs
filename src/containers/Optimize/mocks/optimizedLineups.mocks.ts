/* eslint-disable sort-keys */
export const optimizedYahoolineups = [
	{
		players: [
			{
				firstName: 'Jacob',
				fppg: 35.21,
				id: 3592794,
				image: 'https://s.yimg.com/bt/api/res/1.2/I5jIg3zecS1HNn0ePEbQYw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05042022/9701.png',
				lastName: 'deGrom',
				position: 'P',
				salary: 58,
				status: 'IL60',
				team: 'NYM',
			},
			{
				firstName: 'Tyler',
				fppg: 28.87,
				id: 3592426,
				image: 'https://s.yimg.com/bt/api/res/1.2/G1jRy9TqRdsP2BOkOpiWRA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/06042022/9616.png',
				lastName: 'Glasnow',
				position: 'P',
				salary: 46,
				status: 'IL60',
				team: 'TB',
			},
			{
				firstName: 'William',
				fppg: 7.57,
				id: 3592480,
				image: 'https://s.yimg.com/bt/api/res/1.2/xg1xk_CrvY70UpYkMm6P1w--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/04202022/11268.png',
				lastName: 'Contreras',
				position: 'C',
				salary: 9,
				status: 'N/A',
				team: 'ATL',
			},
			{
				firstName: 'Paul',
				fppg: 10.72,
				id: 3592527,
				image: 'https://s.yimg.com/bt/api/res/1.2/qob2EwBxCh9jbM1yuFIcNw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05022022/8967.png',
				lastName: 'Goldschmidt',
				position: '1B',
				salary: 7,
				status: 'DTD',
				team: 'STL',
			},
			{
				firstName: 'Nick',
				fppg: 17.4,
				id: 3592521,
				image: 'https://s.yimg.com/bt/api/res/1.2/smLFC3qT7wGtBCwMkAB.Hw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05132022/11814.png',
				lastName: 'Maton',
				position: '2B',
				salary: 9,
				status: 'IL10',
				team: 'PHI',
			},
			{
				firstName: 'Rafael',
				fppg: 9.77,
				id: 3592646,
				image: 'https://s.yimg.com/bt/api/res/1.2/jRP34h5MB55a2VEWteO7dA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05022022/10235.png',
				lastName: 'Devers',
				position: '3B',
				salary: 7,
				status: 'IL10',
				team: 'BOS',
			},
			{
				firstName: 'Fernando',
				fppg: 10.71,
				id: 3592692,
				image: 'https://s.yimg.com/bt/api/res/1.2/yHVrxXRFkQKscGkS65rIdA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/06072021/10639.png',
				lastName: 'Tatis Jr.',
				position: 'SS',
				salary: 7,
				status: 'IL60',
				team: 'SD',
			},
			{
				firstName: 'Aaron',
				fppg: 11.31,
				id: 3592741,
				image: 'https://s.yimg.com/bt/api/res/1.2/dkT7NQ28lBgDfbXlB.q.TQ--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/04282022/9877.png',
				lastName: 'Judge',
				position: 'OF',
				salary: 25,
				status: 'N/A',
				team: 'NYY',
			},
			{
				firstName: 'Bryce',
				fppg: 10.54,
				id: 3592519,
				image: 'https://s.yimg.com/bt/api/res/1.2/Qi1QmoSYRwNnYqKJk22wSA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05132022/8875.png',
				lastName: 'Harper',
				position: 'OF',
				salary: 7,
				status: 'IL10',
				team: 'PHI',
			},
			{
				firstName: 'Shohei',
				fppg: 11.77,
				id: 3592808,
				image: 'https://s.yimg.com/bt/api/res/1.2/HfkPzSVZdTfgai.Zx5YFlw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05122022/10835.png',
				lastName: 'Ohtani',
				position: 'OF',
				salary: 18,
				status: 'DTD',
				team: 'LAA',
			},
		],
		salary: 153.87,
		fppg: 193,
	},
	{
		players: [
			{
				firstName: 'Jacob',
				fppg: 35.21,
				id: 3592794,
				image: 'https://s.yimg.com/bt/api/res/1.2/I5jIg3zecS1HNn0ePEbQYw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05042022/9701.png',
				lastName: 'deGrom',
				position: 'P',
				salary: 58,
				status: 'IL60',
				team: 'NYM',
			},
			{
				firstName: 'Tyler',
				fppg: 28.87,
				id: 3592426,
				image: 'https://s.yimg.com/bt/api/res/1.2/G1jRy9TqRdsP2BOkOpiWRA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/06042022/9616.png',
				lastName: 'Glasnow',
				position: 'P',
				salary: 46,
				status: 'IL60',
				team: 'TB',
			},
			null,
			{
				firstName: 'Paul',
				fppg: 10.72,
				id: 3592527,
				image: 'https://s.yimg.com/bt/api/res/1.2/qob2EwBxCh9jbM1yuFIcNw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05022022/8967.png',
				lastName: 'Goldschmidt',
				position: '1B',
				salary: 7,
				status: 'DTD',
				team: 'STL',
			},
			{
				firstName: 'Nick',
				fppg: 17.4,
				id: 3592521,
				image: 'https://s.yimg.com/bt/api/res/1.2/smLFC3qT7wGtBCwMkAB.Hw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05132022/11814.png',
				lastName: 'Maton',
				position: '2B',
				salary: 9,
				status: 'IL10',
				team: 'PHI',
			},
			{
				firstName: 'Rafael',
				fppg: 9.77,
				id: 3592646,
				image: 'https://s.yimg.com/bt/api/res/1.2/jRP34h5MB55a2VEWteO7dA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05022022/10235.png',
				lastName: 'Devers',
				position: '3B',
				salary: 7,
				status: 'IL10',
				team: 'BOS',
			},
			{
				firstName: 'Fernando',
				fppg: 10.71,
				id: 3592692,
				image: 'https://s.yimg.com/bt/api/res/1.2/yHVrxXRFkQKscGkS65rIdA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/06072021/10639.png',
				lastName: 'Tatis Jr.',
				position: 'SS',
				salary: 7,
				status: 'IL60',
				team: 'SD',
			},
			{
				firstName: 'Aaron',
				fppg: 11.31,
				id: 3592741,
				image: 'https://s.yimg.com/bt/api/res/1.2/dkT7NQ28lBgDfbXlB.q.TQ--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/04282022/9877.png',
				lastName: 'Judge',
				position: 'OF',
				salary: 25,
				status: 'N/A',
				team: 'NYY',
			},
			{
				firstName: 'Bryce',
				fppg: 10.54,
				id: 3592519,
				image: 'https://s.yimg.com/bt/api/res/1.2/Qi1QmoSYRwNnYqKJk22wSA--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05132022/8875.png',
				lastName: 'Harper',
				position: 'OF',
				salary: 7,
				status: 'IL10',
				team: 'PHI',
			},
			{
				firstName: 'Shohei',
				fppg: 11.77,
				id: 3592808,
				image: 'https://s.yimg.com/bt/api/res/1.2/HfkPzSVZdTfgai.Zx5YFlw--/YXBwaWQ9ZGFpbHlmYW50YXN5O3E9ODU7Y2M9ODY0MDA7dz0xMjA7aD0xMjA-/https://s.yimg.com/xe/i/us/sp/v/mlb_cutout/players_l/05122022/10835.png',
				lastName: 'Ohtani',
				position: 'OF',
				salary: 18,
				status: 'DTD',
				team: 'LAA',
			},
		],
		salary: 153.57,
		fppg: 194,
	},
];
