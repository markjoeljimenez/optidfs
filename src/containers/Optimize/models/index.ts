export * from './IGetOptimizedLineupsPostBody';
export * from './IGetOptimizedLineupsResponse';
export * from './IOptimizedLineup';
export * from './IOptimizeSettings';
