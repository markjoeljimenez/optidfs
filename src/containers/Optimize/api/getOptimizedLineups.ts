import {
	BaseQueryFn,
	FetchArgs,
	FetchBaseQueryError,
	FetchBaseQueryMeta,
} from '@reduxjs/toolkit/dist/query';
import { EndpointBuilder } from '@reduxjs/toolkit/dist/query/endpointDefinitions';

import { IPlayer } from '@/containers/Players';
import { EProviders } from '@/containers/Providers';
import { ISport } from '@/containers/Sports';

import { IGetOptimizedLineupsResponse, IOptimizeSettings } from '../models';

export const getOptimizedLineups = (
	builder: EndpointBuilder<
		BaseQueryFn<
			string | FetchArgs,
			unknown,
			FetchBaseQueryError,
			{},
			FetchBaseQueryMeta
		>,
		never,
		'optidfs'
	>
) =>
	builder.mutation<
		IGetOptimizedLineupsResponse,
		{
			players: IPlayer[];
			provider: EProviders;
			settings: IOptimizeSettings;
			sport: ISport;
		}
	>({
		query: (body) => {
			return {
				body,
				method: 'POST',
				url: '/optimize',
			};
		},
	});
