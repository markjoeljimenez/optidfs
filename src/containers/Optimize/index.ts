import { getOptimizedLineups } from './api/getOptimizedLineups';
import Optimize from './components/Optimize.component';
import optimizeHandler from './handlers/optimize.handler';
import useGetOptimizedLineupsMutationResponse from './hooks/useGetOptimizedLineupsMutationResponse';
import IGetOptimizedLineupsResponse from './models/IGetOptimizedLineupsResponse';
import IOptimizedLineup from './models/IOptimizedLineup';
import OptimizeReducers, { optimizedState } from './redux/Optimize.reducers';

// Api
export { getOptimizedLineups };

// Handlers
export { optimizeHandler };

// Reducers
export { OptimizeReducers };

// State
export { optimizedState };

// Models
export type { IGetOptimizedLineupsResponse, IOptimizedLineup };

// Hooks
export { useGetOptimizedLineupsMutationResponse };

// Components
export default Optimize;
