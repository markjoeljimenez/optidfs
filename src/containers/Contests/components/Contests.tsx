import { IContest } from '../interfaces/IContest';
import { IDraftKingsContest } from '../interfaces/IDraftKingsContest';
import { IYahooContest } from '../interfaces/IYahooContest';
import { selectProviders } from '@/containers/Providers';
import { setGameType, setSelectedContest } from '../redux/reducers';
import { sportsState } from '@/containers/Sports';
import { useAppDispatch, useAppSelector } from '../../../hooks';
import { useCombobox, UseComboboxStateChange } from 'downshift';
import { useGetContestsFromSportQuery } from '../../../api';
import { useState, useEffect, useRef } from 'react';
import {
	mapDraftKingsContests,
	mapYahooContests,
} from '../services/mapContests';
import clsx from 'clsx';
import { useLocalStorage } from 'react-use';

const Dropdown = () => {
	const { sports, providers, contests } = useAppSelector((state) => state);
	// const sports = useAppSelector(sportsState);
	// const providers = useAppSelector(selectProviders);
	const dispatch = useAppDispatch();

	const [localStorage, setLocalStorage] = useLocalStorage(
		'optidfs-initial-visit'
	);

	const { data } = useGetContestsFromSportQuery({
		sportId: sports.selectedSport!.sportId,
		sport: sports.selectedSport!.regionAbbreviatedSportName,
		provider: providers.provider!,
	});

	const [defaultContests, setDefaultContests] = useState<IContest[]>([]);
	const [filteredContests, setFilteredContests] = useState<IContest[]>([]);

	function onStateChange(selection: UseComboboxStateChange<IContest>) {
		if (!selection?.selectedItem) {
			return;
		}

		const { selectedItem } = selection;

		if (selectedItem) {
			dispatch(setSelectedContest(selectedItem));

			if (selectedItem.gameType) {
				dispatch(setGameType(selectedItem.gameType));
			}

			// dispatch(setContest(selectedItem));
			// dispatch(getPlayers(selectedItem.id));
		}
	}

	useEffect(() => {
		if (data) {
			const transformedContests =
				data.provider === 'draftkings'
					? mapDraftKingsContests(
							data.contests as IDraftKingsContest[]
					  )
					: mapYahooContests(data.contests as IYahooContest[]);

			setDefaultContests(transformedContests);
			setFilteredContests(transformedContests);
		}
	}, [data, localStorage]);

	const {
		isOpen,
		getToggleButtonProps,
		getLabelProps,
		getMenuProps,
		getInputProps,
		getComboboxProps,
		getItemProps,
	} = useCombobox({
		itemToString: (item) => (item ? item.name : ''),
		items: filteredContests,
		onInputValueChange: ({ inputValue }) => {
			setFilteredContests(
				inputValue && inputValue !== ''
					? defaultContests.filter((contest) =>
							contest.name
								.toLowerCase()
								.includes(inputValue?.toLocaleLowerCase())
					  )
					: defaultContests
			);
		},
		selectedItem: defaultContests.find(
			(contest) => contest.id === contests?.selectedContest?.id
		),
		onStateChange,
	});

	return (
		<div className="relative" {...getComboboxProps()}>
			{/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
			<label
				className="hidden"
				htmlFor="select-contest"
				{...getLabelProps()}
			>
				Search contest by ID or name
			</label>
			<input
				id="select-contest"
				className="md:min-w-dropdown h-full shadow appearance-none border rounded w-full py-3 px-4 pr-12 text-gray-700 leading-tight focus:outline-none focus:shadow-outline truncate"
				{...getInputProps({
					placeholder: 'Search contest by ID or name',
				})}
			/>
			<button
				{...getToggleButtonProps()}
				type="button"
				className="absolute inset-y-0 right-0 px-2 text-gray-700 border-l border-gray-300"
			>
				<svg
					className="fill-current"
					xmlns="http://www.w3.org/2000/svg"
					viewBox="0 0 24 24"
					width="24"
					height="24"
				>
					<g data-name="Layer 2">
						<g data-name="chevron-down">
							<rect width="24" height="24" opacity="0" />
							<path d="M12 15.5a1 1 0 0 1-.71-.29l-4-4a1 1 0 1 1 1.42-1.42L12 13.1l3.3-3.18a1 1 0 1 1 1.38 1.44l-4 3.86a1 1 0 0 1-.68.28z" />
						</g>
					</g>
				</svg>
				<span className="sr-only">Down</span>
			</button>
			<ul
				className={clsx(
					isOpen &&
						'absolute top-1/1 left-0 right-0 max-h-20 bg-white overflow-y-scroll shadow border rounded mt-4 z-10'
				)}
				{...getMenuProps()}
			>
				{isOpen && filteredContests?.length ? (
					<>
						{filteredContests?.map((item, index) => (
							<li
								className="p-4 border-b border-gray-300 hover:bg-gray-100 cursor-pointer"
								{...getItemProps({
									index,
									item,
								})}
								key={index}
							>
								{item.gameType || item.id} - {item.name}
							</li>
						))}
					</>
				) : null}
			</ul>
		</div>
	);
};

export default Dropdown;
