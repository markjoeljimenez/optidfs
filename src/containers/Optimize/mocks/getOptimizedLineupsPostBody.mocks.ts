import { mapYahooPlayers, yahooPlayersMock } from '@/containers/Players';
import { EProviders } from '@/containers/Providers';
import { yahooSportsMock } from '@/containers/Sports';

import { IGetOptimizedLineupsPostBody } from '../models';

export const GetOptimizedLineupsPostBodyMock: IGetOptimizedLineupsPostBody = {
	players: mapYahooPlayers(yahooPlayersMock),
	provider: EProviders.Yahoo,
	settings: {
		excludedPlayers: [],
		lockedPlayers: [],
		numberOfLineups: 1,
		statusFilters: ['N/A', 'DTD'],
	},
	sport: yahooSportsMock[1],
};
