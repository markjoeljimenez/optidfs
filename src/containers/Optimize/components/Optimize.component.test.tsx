import userEvent from '@testing-library/user-event';
import { RootState } from 'src/store';

import RightNavigation from '@/components/global/right-navigation';
import { mapContests, yahooContestsMock } from '@/containers/Contests';
import { GetPlayersExtendedApi, IPlayer } from '@/containers/Players';
import { EProviders } from '@/containers/Providers';
import { yahooSportsMock } from '@/containers/Sports';
import { render, screen, waitFor } from '@/test/render';

import { GetOptimizedLineupsExtendedApi } from '../api';
import { GetOptimizedLineupsPostBodyMock } from '../mocks';
import { Optimize } from './Optimize.component';

const preloadedState: Partial<RootState> = {
	contests: {
		gameType: null,
		selectedContest: mapContests(yahooContestsMock, 'yahoo')[0],
	},
	optimize: GetOptimizedLineupsPostBodyMock,
	providers: {
		provider: EProviders.Yahoo,
	},
	sports: {
		selectedSport: yahooSportsMock[0],
	},
};

const app = (
	<>
		<Optimize />
		<RightNavigation />
	</>
);

jest.mock('flagsmith/react', () => ({
	useFlags: jest.fn().mockReturnValue({
		stacking: {
			enabled: true,
			value: null,
		},
	}),
}));

describe('Optimize', () => {
	const getPlayers = GetPlayersExtendedApi.endpoints.getPlayers.select({
		id: preloadedState.contests?.selectedContest?.contest_id!,
		provider: EProviders.Yahoo,
	});
	const getOptimizedLineups =
		GetOptimizedLineupsExtendedApi.endpoints.getOptimizedLineups.select(
			'optimize'
		);

	it('should match snapshot', async () => {
		// Arrange
		const view = render(app, { preloadedState });
		await waitFor(() =>
			expect(getPlayers(view.store.getState()).isSuccess).toBe(true)
		);

		// Assert
		expect(view).toMatchSnapshot();
	});

	it('should get optimized lineups successfully', async () => {
		// Arrange
		const { store } = render(app, { preloadedState });
		await waitFor(() =>
			expect(getPlayers(store.getState()).isSuccess).toBe(true)
		);

		// Act
		const optimizeButton = screen.getByTestId('optimize');
		await userEvent.click(optimizeButton);

		await waitFor(() =>
			expect(getOptimizedLineups(store.getState()).isSuccess).toBe(true)
		);

		const getOptimizedLineupsState = getOptimizedLineups(store.getState());

		// Assert
		expect(getOptimizedLineupsState.data?.lineups.length).toBe(1);
	});

	it('should get number of generated lineups if set', async () => {
		// Arrange
		const { store } = render(app, { preloadedState });
		await waitFor(() =>
			expect(getPlayers(store.getState()).isSuccess).toBe(true)
		);

		// Act
		const settingsButtons = screen.getAllByRole('button');
		await userEvent.click(settingsButtons[1]);

		const numberOfGenerationsInput = screen.getByTestId(
			'number-of-generations'
		);
		await userEvent.type(numberOfGenerationsInput, '{Backspace}{5}');

		const optimizeButton = screen.getByTestId('optimize');
		await userEvent.click(optimizeButton);

		await waitFor(() =>
			expect(getOptimizedLineups(store.getState()).isSuccess).toBe(true)
		);

		const getOptimizedLineupsState = getOptimizedLineups(store.getState());

		// Assert
		expect(getOptimizedLineupsState.data?.lineups.length).toBe(5);
	});

	it('should filter players if set', async () => {
		// Arrange
		const { store } = render(app, { preloadedState });
		await waitFor(() =>
			expect(getPlayers(store.getState()).isSuccess).toBe(true)
		);

		// Act
		const settingsButtons = screen.getAllByRole('button');
		await userEvent.click(settingsButtons[1]);

		const filterCheckboxes = screen.getAllByRole('checkbox');
		await userEvent.click(filterCheckboxes[1]); // Check Active
		await userEvent.click(filterCheckboxes[2]); // Check IL10

		const optimizeButton = screen.getByTestId('optimize');
		await userEvent.click(optimizeButton);

		await waitFor(() =>
			expect(getOptimizedLineups(store.getState()).isSuccess).toBe(true)
		);

		const getOptimizedLineupsState = getOptimizedLineups(store.getState());

		// Assert
		expect(
			getOptimizedLineupsState.data!.lineups[0].players.every(
				(player: IPlayer) =>
					player.status === 'N/A' || player.status === 'IL10'
			)
		).toBe(true);
	});
});
