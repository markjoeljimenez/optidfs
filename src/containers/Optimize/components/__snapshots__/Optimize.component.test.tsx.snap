// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Optimize should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="inline-flex items-center px-4 py-2 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
        data-testid="optimize"
        type="button"
      >
        Optimize
      </button>
      <div
        class="bg-white flex sticky right-0"
      >
        <div
          class="py-8 px-4 space-y-4"
        >
          <button
            class="block text-gray-400 hover:text-gray-600"
            value="settings"
          >
            <svg
              class="icon icon-tabler icon-tabler-settings"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
                stroke="none"
              />
              <path
                d="M10.325 4.317c.426 -1.756 2.924 -1.756 3.35 0a1.724 1.724 0 0 0 2.573 1.066c1.543 -.94 3.31 .826 2.37 2.37a1.724 1.724 0 0 0 1.065 2.572c1.756 .426 1.756 2.924 0 3.35a1.724 1.724 0 0 0 -1.066 2.573c.94 1.543 -.826 3.31 -2.37 2.37a1.724 1.724 0 0 0 -2.572 1.065c-.426 1.756 -2.924 1.756 -3.35 0a1.724 1.724 0 0 0 -2.573 -1.066c-1.543 .94 -3.31 -.826 -2.37 -2.37a1.724 1.724 0 0 0 -1.065 -2.572c-1.756 -.426 -1.756 -2.924 0 -3.35a1.724 1.724 0 0 0 1.066 -2.573c-.94 -1.543 .826 -3.31 2.37 -2.37c1 .608 2.296 .07 2.572 -1.065z"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
          </button>
          <button
            class="block text-gray-400 hover:text-gray-600"
            value="stacking"
          >
            <svg
              class="icon icon-tabler icon-tabler-stack-2"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
                stroke="none"
              />
              <polyline
                points="12 4 4 8 12 12 20 8 12 4"
              />
              <polyline
                points="4 12 12 16 20 12"
              />
              <polyline
                points="4 16 12 20 20 16"
              />
            </svg>
          </button>
        </div>
        <div
          class="border-l-[1px] border-gray-200 min-w-[20rem]"
          hidden=""
        >
          <div
            class="p-8"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <button
      class="inline-flex items-center px-4 py-2 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
      data-testid="optimize"
      type="button"
    >
      Optimize
    </button>
    <div
      class="bg-white flex sticky right-0"
    >
      <div
        class="py-8 px-4 space-y-4"
      >
        <button
          class="block text-gray-400 hover:text-gray-600"
          value="settings"
        >
          <svg
            class="icon icon-tabler icon-tabler-settings"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <path
              d="M10.325 4.317c.426 -1.756 2.924 -1.756 3.35 0a1.724 1.724 0 0 0 2.573 1.066c1.543 -.94 3.31 .826 2.37 2.37a1.724 1.724 0 0 0 1.065 2.572c1.756 .426 1.756 2.924 0 3.35a1.724 1.724 0 0 0 -1.066 2.573c.94 1.543 -.826 3.31 -2.37 2.37a1.724 1.724 0 0 0 -2.572 1.065c-.426 1.756 -2.924 1.756 -3.35 0a1.724 1.724 0 0 0 -2.573 -1.066c-1.543 .94 -3.31 -.826 -2.37 -2.37a1.724 1.724 0 0 0 -1.065 -2.572c-1.756 -.426 -1.756 -2.924 0 -3.35a1.724 1.724 0 0 0 1.066 -2.573c-.94 -1.543 .826 -3.31 2.37 -2.37c1 .608 2.296 .07 2.572 -1.065z"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </button>
        <button
          class="block text-gray-400 hover:text-gray-600"
          value="stacking"
        >
          <svg
            class="icon icon-tabler icon-tabler-stack-2"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <polyline
              points="12 4 4 8 12 12 20 8 12 4"
            />
            <polyline
              points="4 12 12 16 20 12"
            />
            <polyline
              points="4 16 12 20 20 16"
            />
          </svg>
        </button>
      </div>
      <div
        class="border-l-[1px] border-gray-200 min-w-[20rem]"
        hidden=""
      >
        <div
          class="p-8"
        >
          <div />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
